Add support for broadcast ops in MIXR

jungpark-mlir opened the pull request:
contains
- migraphx.broadcast
- migraphx.multibroadcast
- fix broadcast folding with nchw layout

jungpark-mlir committed:
Add broadcast lowering

jungpark-mlir committed:
add  multibroadcast lowering

jungpark-mlir committed:
Fix bcast folding layout support

jungpark-mlir committed:
add test

sjw36 commented on mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp:
```diff
@@ -135,9 +135,131 @@ class ConvConverter final
+    for (auto &use : op->getResult(0).getUses()) {
+      auto expandedOp = use.getOwner();
+      // isa binary operation,
+      if (isa<migraphx::AddOp>(expandedOp) ||
```
Shouldn't need to restrict to these ops. Can we just replace the migraphx.broadcast with the tosa.reshape?

jungpark-mlir replied:
broadcast in tosa is restrictive. Just checked and it only broadcasts on the second input of the binary operations. Not clearly defined as a rule but only defined per each operator.
I'll change it to reflect that rule.

jungpark-mlir committed:
Correct broadcast condition

jungpark-mlir committed:
Fix

jungpark-mlir committed:
Enhance the code per review

Don't duplicate reshape for broadcast
swap input0 and 1 if 0 is broadcast
don't reshape when already with the same rank

jungpark-mlir committed:
format

jungpark-mlir committed:
fix

sjw36 commented on mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp:
```diff
@@ -135,9 +147,134 @@ class ConvConverter final
+    auto inShape = input_t.getType().cast<ShapedType>().getShape();
+
+    Value newOperand = input_t;
+    if (outRank != inShape.size()) {
```
Looks good.

sjw36 reviewed (approved):
LGTM
